### PR TITLE
[SC-640] Bug(64): Make asserter pay for the bond

### PR DESCRIPTION
**Pull request by 0xNuggan**
Ensure that the `asserter` actually pays for the bond. Also added a specific error message and updated tests

### CODE REVIEW

File: src/modules/logicModule/abstracts/oracleIntegrations/UMA_OptimisticOracleV3/OptimisticOracleIntegrator.sol [hunk 1]
```diff
@@ -179,9 +179,29 @@ abstract contract OptimisticOracleIntegrator is
...
-        );
+        if (asserter == address(this)) {
+            //ensure we have enough balance
+            if (defaultCurrency.balanceOf(address(this)) < defaultBond) {
+                revert
+                    Module__OptimisticOracleIntegrator_InsufficientFundsToPayForBond(
+                );
+            }
+        } else {
+            (bool success, bytes memory data) = address(defaultCurrency).call(
```
Comment (Zitzak) [inline src/modules/logicModule/abstracts/oracleIntegrations/UMA_OptimisticOracleV3/OptimisticOracleIntegrator.sol:203]: `data` shadows exisiting declaration. Consider making the functiona parameter `_data`

File: test/modules/logicModule/oracle/OptimisticOracleIntegrator.t.sol [hunk 2]
```diff
@@ -301,20 +301,19 @@ contract OptimisticOracleIntegratorTest is ModuleTest {
...
                         emits an event
 
 
     */
-    // Maybe better here?
-    function testAssertDataFor_whenCallerDoesNotHaveAsserterRole(address who)
-        public
-    {
+    function testAssertDataForFails_whenCallerDoesNotHaveAsserterRole(
+        address who
```
Comment (Zitzak) [inline test/modules/logicModule/oracle/OptimisticOracleIntegrator.t.sol]: unused variable

**Review by Zitzak — CHANGES_REQUESTED** [review]
Not to familar with the OOI so good that @marvinkruse  looks at it as well. Found 2 small things

File: src/modules/logicModule/abstracts/oracleIntegrations/UMA_OptimisticOracleV3/OptimisticOracleIntegrator.sol (hunk 1 above)
Comment (Zitzak) [reply src/modules/logicModule/abstracts/oracleIntegrations/UMA_OptimisticOracleV3/OptimisticOracleIntegrator.sol:203]: variable was actually contained within the else code block. Still updated it for readability and pushed the changes. Waiting for @marvinkruse review to finalize

File: src/modules/logicModule/abstracts/oracleIntegrations/UMA_OptimisticOracleV3/OptimisticOracleIntegrator.sol [hunk 3]
```diff
@@ -179,9 +179,29 @@ abstract contract OptimisticOracleIntegrator is
         returns (bytes32 assertionId)
     {
         asserter = asserter == address(0) ? _msgSender() : asserter;
-        defaultCurrency.safeTransferFrom(
-            _msgSender(), address(this), defaultBond
-        );
+        if (asserter == address(this)) {
+            //ensure we have enough balance
```
Comment (marvinkruse) [inline src/modules/logicModule/abstracts/oracleIntegrations/UMA_OptimisticOracleV3/OptimisticOracleIntegrator.sol]: Whitespace between `//` and `ensure` is missing.

File: src/modules/logicModule/abstracts/oracleIntegrations/UMA_OptimisticOracleV3/OptimisticOracleIntegrator.sol [hunk 4]
```diff
@@ -179,9 +179,29 @@ abstract contract OptimisticOracleIntegrator is
...
+        } else {
+            (bool success, bytes memory data) = address(defaultCurrency).call(
+                abi.encodeWithSelector(
+                    defaultCurrency.transferFrom.selector,
+                    _msgSender(),
+                    address(this),
+                    defaultBond
+                )
+            );
+            //require(success && (data.length == 0 || abi.decode(data, (bool)))); -> taken over from SafeERC20 since we want to override the revert messsage
```
Comment (marvinkruse) [inline src/modules/logicModule/abstracts/oracleIntegrations/UMA_OptimisticOracleV3/OptimisticOracleIntegrator.sol]: Whitespace between `//` and `require` is missing.

**Comment by Zitzak** [comment]
Added the change myself quickly @0xNuggan, jfyi

**Review by Zitzak — APPROVED** [review]

**Review by marvinkruse — CHANGES_REQUESTED** [review]
Otherwise good!

**Comment by 0xNuggan** [comment]
Thanks for the reviews! Fixed the whitespace

**Review by marvinkruse — APPROVED** [review]
good work, gents 🙏

**Comment by 0xNuggan** [comment]
While fixing bug 118 I noticed an edge case that's not covered here. Will update and reopen once I'm done there

**Comment by 0xNuggan** [comment]
The case not covered was if the defaultBond token is not actually a contract. The added fix should address that

**Review by Zitzak — APPROVED** [review]
LGTM.

If we want to add additional check for it to be a token we could add a balance check as implemented here, not sure if needed here though
[[redo-sc-633-bug55-if-invertertransparentupgradeableproxy_v1-wont-not-be](https://github.com/InverterNetwork/contracts/blob/72210f82338e7170990f5287aed159e0c83dd87c/src/modules/authorizer/role/AUT_TokenGated_Roles_v1.sol#L135)](https://github.com/InverterNetwork/contracts/blob/72210f82338e7170990f5287aed159e0c83dd87c/src/modules/authorizer/role/AUT_TokenGated_Roles_v1.sol#L135)

**Comment by 0xNuggan** [comment]
Yes! I do that in PR #600 for validation of paymentOrders() since I felt it was out of scope for this PR :)

**Pull request merged**